p/a/path: fail on broken pattern early, add tests (#250)
brancz/kube-rbac-proxy: docs/path.md

> # Allow Path and Ignore Path

Reviewer: you may want to create a separate `docs/migration` folder for these files

## Description

If `--allow-paths` is set, only the requests to these paths will be considered for proxying, the rest will be rejected.
If `--ignore-paths` is set, requests to these paths will be passed upstream without any authorization.

Only one of `--allow-paths`/`--ignore-paths` is allowed to be set at a time.

## Change

Previously, it was possible to configure an **infix wildcard** but that will now fail.
Now **infix regex** will cause an error on server startup.
kube-rbac-proxy won't start with an **infix regex**.
E.g. `--allow-path="/api/v1/*/values"` will cause kube-rbac-proxy fail to start.
E.g. `--ignore-path="/api/v1/*/values"` will cause kube-rbac-proxy fail to start.

Previously `*` would count as a single-path-segment wildcard ([path.Match](https://pkg.go.dev/path#Match)).
Now the wildcard is matching any string, even if it contains `/`.

> E.g. `--allow-path="/api/v1/*"` would have rejected `/api/v1/label/values`, not it evaluated.

Reviewer: `not it evaluated` <- that sentence appears to be missing something

> E.g. `--ignore-path="/api/v1/*"` would have evaluated `/api/v1/label/values`, now it is passed through. 

Reviewer: ```suggestion
E.g. `--ignore-path="/api/v1/*"` would have matched `/api/v1/label/values`, now it is passed through. 
```

# Allow Path and Ignore Path

## Description

> Paths that are not provided by `--allow-path` will be rejected.
> Paths that are provided by `--ignore-path` won't be rejected.
> It is only possible to define one of both.

Reviewer: I think this is a bit confusing. How about:

```suggestion
If `--allow-paths` is set, only the requests to these paths will be considered for proxying, the rest will be rejected.
If `--ignore-paths` is set, requests to these paths will be passed upstream without any authorization.

Only one of `--allow-paths`/`--ignore-paths` is allowed to be set at a time.
```

Note that the README says `allow/ignore-paths` whereas here you're describing it as `allow/ignore-path`

> ## Change

Reviewer: Should we have a separate section for changes to previous kube-rbac-versions?

Author: Do you mean in addition to the CHANGELOG.md?

> Previously **infix regex** worked.

Reviewer: ```suggestion
Previously, it was possible to configure an **infix wildcard** but that will now fail.
```

Now **infix regex** will cause an error on server startup.
kube-rbac-proxy won't start with an **infix regex**.
E.g. `--allow-path="/api/v1/*/values"` will cause kube-rbac-proxy fail to start.
E.g. `--ignore-path="/api/v1/*/values"` will cause kube-rbac-proxy fail to start.

> Previously `*` would count as a wildcard in between `/`.

Reviewer: Wasn't it a single-path-segment wildcard?

Now the wildcard is matching any string, even if it contains `/`.
E.g. `--allow-path="/api/v1/*"` would have rejected `/api/v1/label/values`, not it evaluated.
E.g. `--ignore-path="/api/v1/*"` would have evaluated `/api/v1/label/values`, now it is passed through. 

> Only `*` is allowed. Previously everything that [Go's standard library path.Match](https://pkg.go.dev/path#Match) allowed worked.

Reviewer: ```suggestion
Only a single `*` is allowed. Previously everything that [Go's standard library path.Match](https://pkg.go.dev/path#Match) allowed worked.
```
